### PR TITLE
Feat subtitle plugin

### LINKED ISSUE #2976: Plugin to record subtitles for video
With a video recording enabled for Playwright it is important also to show the progress of a test within that video.
The easiest solution so far is to generate subtitles to each video recorded by Playwright.

* Subtitles will include all executed steps and saved into *.srt file. 
* This srt file should be named after a video file which was saved (can be obtained from `test.artifacts.video` **after** a test)

The flow is:

* create a plugin
* listen to all `step.before` events
* write all steps into memory
* after a test check for `test.artifacts.video` to be found
* if this is found - write subtitles file

### PR DESCRIPTION
**Pull request by anirudh-modi**
## Motivation/Description of the PR
- This PR is to capture steps of the tests as a subtitle and then report them as an artifact if a video was generated was the corresponding failing test
- Resolves #2976 

Applicable helpers:

- [ ] WebDriver
- [ ] Puppeteer
- [ ] Nightmare
- [ ] REST
- [ ] FileHelper
- [ ] Appium
- [ ] Protractor
- [ ] TestCafe
- [x] Playwright

Applicable plugins:

- [ ] allure
- [ ] autoDelay
- [ ] autoLogin
- [ ] customLocator
- [ ] pauseOnFail
- [ ] puppeteerCoverage
- [ ] retryFailedStep
- [ ] screenshotOnFail
- [ ] selenoid
- [ ] stepByStepReport
- [ ] wdio

## Type of change

- [ ] :fire: Breaking changes
- [x] :rocket: New functionality
- [ ] :bug: Bug fix
- [ ] :clipboard: Documentation changes/updates
- [ ] :hotsprings: Hot fix
- [ ] :hammer: Markdown files fix - not related to source code
- [ ] :nail_care: Polish code

## Checklist:

- [x] Tests have been added
- [x] Documentation has been added (Run `npm run docs`)
- [x] Lint checking (Run `npm run lint`)
- [x] Local tests are passed (Run `npm test`)

### CODE REVIEW

File: docs/plugins.md [hunk 1]
```diff
@@ -306,6 +306,109 @@ Scenario('login', async ( {I, login} ) => {
 
 -   `config`  
 
+## captureStepAsSubtitle
```
Comment (anirudh-modi) [inline docs/plugins.md:309]: Please ignore this file as this is a WIP

File: lib/plugin/captureStepAsSubtitle.js [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+ *   When();
+ *   projectPage.update(projectId, { title: 'new title' });
+ *
+ *   Then();
+ *   projectPage.open(projectId);
+ *   I.see('new title', 'h1');
+ * });
+ * ```
+ */
+module.exports = function (config) {
```
Comment (anirudh-modi) [inline lib/plugin/captureStepAsSubtitle.js:104]: Any feedback on this file would be very helpful

**Comment by anirudh-modi** [comment]
@DavertMik feedbacks are much welcome..

File: lib/plugin/captureStepAsSubtitle.js [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+  event.dispatcher.on(event.test.before, (test) => {
+    testStartedAt = Date.now();
+    steps = {};
```
Comment (anirudh-modi) [inline lib/plugin/captureStepAsSubtitle.js:107]: This code will mostly break when ran in parallel, however, as an alternative I can maintain a `key...value` pair for each test by its `id` and its corresponding steps..

But, is there any alternative for the same?

File: package.json [hunk 4]
```diff
@@ -84,7 +84,9 @@
     "promise-retry": "^1.1.1",
     "requireg": "^0.2.2",
     "resq": "^1.10.0",
-    "sprintf-js": "^1.1.1"
+    "sprintf-js": "^1.1.1",
+    "subtitle": "^4.1.0",
```
Comment (DavertMik) [inline package.json:88]: do you really need this package?

File: lib/plugin/captureStepAsSubtitle.js (hunk 3 above)
Comment (DavertMik) [reply lib/plugin/captureStepAsSubtitle.js:107]: 1. No, it won't break in parallel. All tests are running in isolated workers
2. seems fine

**Review by DavertMik — APPROVED** [review]

File: lib/plugin/captureStepAsSubtitle.js [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+
+      sorted.forEach((step, index) => {
+        subtitle = `${subtitle}${index + 1}
+${step.start} --> ${step.end}
+${step.title}
+
+`;
+      });
+      const videoPath = require('path').dirname(test.artifacts.video);
+      const videoFileName = require('path').basename(test.artifacts.video, '.webm');
```
Comment (DavertMik) [inline lib/plugin/captureStepAsSubtitle.js:147]: in theory, an extension can be different....

Comment (anirudh-modi) [reply lib/plugin/captureStepAsSubtitle.js:147]: Will fix this..

**Comment by DavertMik** [comment]
@anirudh-modi 
Everything is fine except the name of the plugin

`captureStepAsSubtitle`

Let's simplify it just to `subtitles`

File: docs/plugins.md [hunk 6]
```diff
@@ -306,6 +306,22 @@ Scenario('login', async ( {I, login} ) => {
...
+
+Automatically captures steps as subtitle, and saves it as an artifact when a video is found for a failed test
```
Comment (DavertMik) [inline docs/plugins.md:311]:
```suggestion
Automatically captures steps as subtitles, and saves then as an artifact when a video is found for a failed test
```

File: docs/plugins.md [hunk 7]
```diff
@@ -306,6 +306,22 @@ Scenario('login', async ( {I, login} ) => {
...
+
+```js
+plugins: {
+ captureStepAsSubtitle: {
+   enabled: true
+ }
+}
+```
+
+`captureStepAsSubtitle` plugin should only be enabled, when `video: true` is enabled for playwright config.
```
Comment (DavertMik) [inline docs/plugins.md:323]: > `captureStepAsSubtitle` plugin should only be enabled, when `video: true` is enabled for playwright config.

Not a correct statement. The plugin can be enabled but it won't record any subtitles...

**Review by DavertMik — APPROVED** [review]
Rename a plugin and I'm merging it!

**Comment by anirudh-modi** [comment]
@DavertMik renamed the plugin as `subtitles` as requested.

**Pull request merged**